### PR TITLE
Support calendar aggregation operations

### LINKED ISSUE #2144: Feature request: Add support for calendar aggregation operations
### Similar requests

- [X] I have searched and found no existing similar requests

### How are you using Legend?

Legend Studio

### What problems are you trying to solve?

Add support for calendar aggregation operations

### Describe the solution you would like to see implemented

Add support for calendar aggregation operations

### Model data

```
###Relational
Database test::store
(
  Table EmployeeTable
  (
    id INTEGER PRIMARY KEY,
    hireDate DATE,
    hireType VARCHAR(10),
    fteFactor DOUBLE
  )
  Table FirmTable
  (
    firmName VARCHAR(100) PRIMARY KEY,
    employeeId INTEGER PRIMARY KEY
  )

  Join JoinEmployeeToFirm(EmployeeTable.id = FirmTable.employeeId)
)


###Service
Service test::service
{
  pattern: '/test/calendar/';
  owners:
  [
    'gayathri'
  ];
  documentation: '';
  autoActivateUpdates: true;
  execution: Single
  {
    query: |test::Employee.all()->groupBy([], [agg(x|$x.hireDate->ytd('NY', %2022-11-16, $x.fteFactor), x|$x->sum())], ['ytd']);
    mapping: test::mapping;
    runtime: test::runtime;
  }
  testSuites:
  [
    testSuite_1:
    {
      data:
      [
      ]
      tests:
      [
        test_1:
        {
          serializationFormat: PURE;
          asserts:
          [
            assertion_1:
              EqualToJson
              #{
                expected : 
                  ExternalFormat
                  #{
                    contentType: 'application/json';
                    data: '';
                  }#;
              }#
          ]
        }
      ]
    }
  ]
}


###Pure
Class test::Employee
{
  id: Integer[1];
  hireDate: Date[1];
  hireType: String[1];
  fteFactor: Float[1];
  firmName: String[0..1];
  endDate: Date[1];
}

Class test::temp
{
  id: Number[1];
}

function test::testYtd(): Any[*]
{
   test::Employee.all()->groupBy([p|$p.hireDate], [agg(p|$p.hireDate->ytd('NY', %2022-11-16, $p.fteFactor), y|$y->sum())], ['includedDate', 'ytd'])
}


###Mapping
Mapping test::mapping
(
  test::Employee[emp]: Relational
  {
    ~primaryKey
    (
      [test::store]EmployeeTable.id
    )
    ~mainTable [test::store]EmployeeTable
    id: [test::store]EmployeeTable.id,
    hireDate: [test::store]EmployeeTable.hireDate,
    hireType: [test::store]EmployeeTable.hireType,
    fteFactor: [test::store]EmployeeTable.fteFactor,
    firmName: [test::store]@JoinEmployeeToFirm | [test::store]FirmTable.firmName,
    endDate: [test::store]EmployeeTable.hireDate
  }
  *test::temp: Pure
  {
    ~src test::temp
    id: $src.id
  }
)


###Connection
RelationalDatabaseConnection test::connection
{
  store: test::store;
  type: H2;
  specification: LocalH2
  {
  };
  auth: DefaultH2;
}


###Runtime
Runtime test::runtime
{
  mappings:
  [
    test::mapping
  ];
  connections:
  [
    test::store:
    [
      connection_1: test::connection
    ]
  ];
}
```

### Documentation, Design, Adoption, Migration Strategy

_No response_

### Contribution

- [x] I would like to work on this feature

### PR DESCRIPTION
**Pull request by gayathrir11**
## Summary

closes #2144 

## How did you test this change?

- [ ] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

***Calendar feature***

https://user-images.githubusercontent.com/87973126/235157480-647ef030-3c9e-4ebe-83db-bb6a94cdcace.mp4

***DnD is not allowed for non-compatible types***

https://user-images.githubusercontent.com/87973126/233660412-9aa3972d-823f-4651-947c-2f02ec512f06.mp4

***Behaviour when we don't specify date column field***

https://user-images.githubusercontent.com/87973126/235155865-85e6e9cd-254c-4cfc-a3d2-e50e174a0f71.mp4

***Toggling of calendar form***

https://user-images.githubusercontent.com/87973126/235156596-5ffec7ec-caf1-4377-a0a6-e6f9af0e3ace.mp4

***Light theme***

light-theme

### CODE REVIEW

File: packages/legend-query-builder/src/components/QueryBuilder.tsx [hunk 1]
```diff
@@ -299,6 +302,54 @@ export const QueryBuilder = observer(
...
+              type: ActionAlertActionType.PROCEED,
+              default: true,
+            },
+          ],
+        });
+      } else {
+        queryBuilderState.applicationStore.alertService.setActionAlertInfo({
+          message:
+            'You have just enabled Calendar Aggregation operations. This will let you add calendar functions to the aggregation operations that you perform on projection columns. Do you want to proceed?',
+          type: ActionAlertType.CAUTION,
```
Comment (akphi) [inline packages/legend-query-builder/src/components/QueryBuilder.tsx:336]: Add a prompt for the question: `prompt: Do you want to proceed?`

File: packages/legend-query-builder/src/components/QueryBuilder.tsx [hunk 2]
```diff
@@ -299,6 +302,54 @@ export const QueryBuilder = observer(
     const openWatermark = (): void => {
       queryBuilderState.watermarkState.setIsEditingWatermark(true);
     };
+
+    const toggleEnableCalendar = (): void => {
+      if (queryBuilderState.isCalendarEnabled) {
+        queryBuilderState.applicationStore.alertService.setActionAlertInfo({
+          message:
+            'You have just disabled Calendar Aggregation operations. This will remove all the calendar aggreagtions you added to the query. Do you want to proceed?',
```
[the remark this answers is not in the document]
Comment (akphi) [reply packages/legend-query-builder/src/components/QueryBuilder.tsx:310]: Also, split `Do you want to proceed?` to be the prompt

**Review by akphi — DISMISSED** [review]
The code looks good to me, but let's do a quick run-through of the interaction. I think by default, we should not `show` the calendar line, as it is confusing, only when the aggregation actually has a calendar function specified, also, I think we can pick the first calendar `NY` as the default even before the calendar function is selected

I also think there should be a `DocumentationLink` button somewhere to link to the documentation for different calendar function

File: packages/legend-query-builder/src/components/QueryBuilder.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (gayathrir11) [reply packages/legend-query-builder/src/components/QueryBuilder.tsx:310]: you mean to the action prompt?

File: packages/legend-query-builder/src/stores/QueryBuilderTextEditorState.ts [hunk 3]
```diff
@@ -169,24 +169,31 @@ export class QueryBuilderTextEditorState extends LambdaEditorState {
...
-      this.setQueryRawLambdaState(new QueryBuilderRawLambdaState(rawLambda));
-    }
-    if (mode === QueryBuilderTextEditorMode.JSON) {
-      this.setLambdaJson(
-        JSON.stringify(
-          pruneSourceInformation(
-            this.queryBuilderState.graphManagerState.graphManager.serializeRawValueSpecification(
-              rawLambda,
+    try {
+      const rawLambda = this.queryBuilderState.buildQuery();
```
[the remark this answers is not in the document]
Comment (gayathrir11) [reply packages/legend-query-builder/src/stores/QueryBuilderTextEditorState.ts:173]: We see this errors when date column field is not specified for calendar functions

**Review by MauricioUyaguari — APPROVED** [review]

**Pull request merged**